[css-anchor-position-1] Account for left scrollbar when calculating anchor position

tuankiet65 opened the pull request:
#### 2b1d3c03ed716d5f0c81989398b8deaaada4db40
<pre>
[css-anchor-position-1] Account for left scrollbar when calculating anchor position
<a href="https://rdar.apple.com/155852237">rdar://155852237</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295964">https://bugs.webkit.org/show_bug.cgi?id=295964</a>

Reviewed by Elika Etemad.

If an anchor is in a RTL containing block (horizontal right-to-left,
or vertical-rl), the scrollbar of the containing block is on the left.
In that case, adjust the calculated position of the anchor to account
for the left scrollbar

Test: imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar.html: Added.
    - Add test to test anchor in scroller with left-hand-side scrollbar
      (e.g in horizontal RTL or vertical-rl writing mode)

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-borders-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar-expected.txt: Added.
    - Add expectation for iOS only. Test is failing in vertical-rl writing mode.
      Tracked at <a href="https://bugs.webkit.org/show_bug.cgi?id=298893.">https://bugs.webkit.org/show_bug.cgi?id=298893.</a>

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):

Canonical link: <a href="https://commits.webkit.org/300018@main">https://commits.webkit.org/300018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e276a744c171130669e69a7fd366e8d33d31ed6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40682 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73066 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a9464144-d545-43e3-add8-2c9c38dc8caf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49261 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91895 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61132 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/efc6317f-0089-4d06-ab5c-047bbe207ab0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108458 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72583 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4ecef6c9-3ed5-49da-815f-45937946279e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32070 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70992 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102547 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130258 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47913 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36405 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100505 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48281 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100407 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25468 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45812 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23860 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44601 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47771 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53484 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47242 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50589 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48926 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -664,6 +664,13 @@ static LayoutUnit computeInsetValue(CSSPropertyID insetPropertyID, CheckedRef<co
+        && selfLogicalAxis == LogicalBoxAxis::Inline;
+    if (shouldFixUpForLeftScrollbar)
+        anchorRange.moveBy(-downcast<RenderBox>(constraints.container()).verticalScrollbarWidth());
+
```
Some questions:
- Should this logic be incorporated into computeAnchorRectRelativeToContainingBlock() directly? That way it would affect both anchor() and position-area.
- Should this also be done for vertical writing modes? E.g. maybe the logic should be `constraints.physicalAxis() == BoxAxis::Horizontal && constraints.container().shouldPlaceVerticalScrollbarOnLeft()`?

tuankiet65 replied:
Yeah `computeAnchorRectRelativeToContainingBlock` would be more appropriate, I was debating whether to put the logic in there or here

tuankiet65 replied:
I simplified the logic and make it more generic (works with anchor function and `position-area`, works with horizontal RTL and vertical-rl)

fantasai commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -498,6 +498,11 @@ LayoutRect AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock(C
             anchorLocation.moveBy(-anchorBox->view().frameView().scrollPositionRespectingCustomFixedPosition());
     }
 
+    if (CheckedPtr containingBox = dynamicDowncast<RenderBox>(containingBlock)) {
```
I believe this doesn't need to be a CheckedPtr, because the containingBlock is a reference so we can rely on it existing already. So it can just be `auto`. (I _think_ that's how this works, maybe check with someone else...)

tuankiet65 replied:
`CheckedPtr` doesn't participate in reference counting (i.e it doesn't extend the lifetime of the containing block), it just makes sure the containing block lives for the entire duration of the `CheckedPtr`. So it's perfectly fine to use it here.

tuankiet65 committed:
[css-anchor-position-1] Account for left scrollbar when calculating anchor position
rdar://155852237
https://bugs.webkit.org/show_bug.cgi?id=295964

Reviewed by Elika Etemad.

If an anchor is in a RTL containing block (horizontal right-to-left,
or vertical-rl), the scrollbar of the containing block is on the left.
In that case, adjust the calculated position of the anchor to account
for the left scrollbar

Test: imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar.html: Added.
    - Add test to test anchor in scroller with left-hand-side scrollbar
      (e.g in horizontal RTL or vertical-rl writing mode)

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-borders-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-in-scroller-with-left-side-scrollbar-expected.txt: Added.
    - Add expectation for iOS only. Test is failing in vertical-rl writing mode.
      Tracked at https://bugs.webkit.org/show_bug.cgi?id=298893.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):

Canonical link: https://commits.webkit.org/300018@main
